### ISSUE TITLE
Design: How to reduce costs running public servers that solve transactions.

### ISSUE THREAD

**Issue by jansegre** [HathorNetwork/hathor-core#67]
Currently we're running multiple expensive GPU machines to mine transactions. Moving forward we need to reduce this cost.

This is the proposed solution:

- Phase 1 (should start as soon as possible to gauge vast.ai's reliability): 
  - [x] Rent GPUs on vast.ai to replace the ones currently use _as is_ (same Docker images, same params, one per API node) will probably need security groups adjustments.
- Phase 2 (_can_ start as soon as possible, has more impact on reliability than cost reduction considering 2 API nodes and 2 GPU instances):
  - [x] Design the tx-mining API
  - [ ] Implement the tx-mining API server (we can probably use our current stratum implementation, and swap the "manager" portions for a job queue filled from HTTP requests)
  - [ ] Implement support for tx-mining API on wallets (maybe only hathor-wallet-lib would be affected)
  - [ ] Deploy tx-mining API using vast.ai GPUs, and the mobile and desktop wallets, while the two (old self-mining API and tx-mining API) are working we would be using the double number of GPU instances than needed
  - [ ] Deprecate and turn off the self-mining API (that mines txs on the API node)
- Phase 3 (somewhat open, we may choose to diverge to changing the hash algo):
  - [ ] Modify cgminer or bmminer for S9 ASIC to mine txs
  - [ ] Find vendors willing to let us run our own firmware (we know at least one)
  - [ ] Incrementally replace GPUs with ASICs (until we have acceptable reliability on ASICs availability)

**Comment by jansegre** [comment]
There are a few ultimate solutions we have already consider.

1. Have a few (preferably distributed and from different vendors for redundancy) ASIC miners mining all transactions for all API nodes (NxM style, each miner can mine for every node, and each node is connected to every miner). Spare hashrate can go to blocks, but transactions must have ultimate priority. This doesn't require any changes to the protocol, but requires changes to firmwares and vendors willing to let us run our custom firmware.
2. Switching the transaction hash algorithm to something ASIC resistant, so we can tune it to the point of common hardware being able to mine txs fast. This both helps clients mine their own transaction, and also enables the use of less expensive servers for mining transactions on public API nodes.
3. Getting rid of tx PoW altogether. This has been brought up before, but it's really soon to talk  about this.

Because 2. and 3. require breaking changes to the protocol, I'll only propose a path to reach 1.

- First we need a modified firmware that:
  - Can mine transactions
  - Prioritizes transactions over blocks
  - Can connect to multiple nodes and mine transactions from all of them
- We can start using (and also testing) this firmware on the public API nodes, this can already alleviate the current situation by only requiring a couple (or even a single) miner for all API nodes.
- Find ASIC renting vendors willing to let us run our modified firmware. This may require adapting our firmware to the hardware available.

Currently there are two big firmware software in use:

- cgminer: the oldest, and the one we are currently using on the GPUs mining transactions.
  - Pros: we already modified it to mine transactions; it could be immediately usable after further modifications.
  - Cons: it is fairly old and the code is a mess; many vendors have a forked source code that is not trivial to port modifications to.
- bmminer: more recent, smaller code base, supports fewer devices (but most of the newer ones)
  - Pros: probably easier to compile and port between different devices
  - Cons: will require more changes, not immediately testable on our current setup, needs more changes

I propose we start with these two fronts:

- modifying cgminer
- looking for a vendor that let us run our own firmware

**Comment by msbrogli** [comment]
Even though solution 1 seems to be a good one, I think it is too complex. As far as I remember, our fork of cgminer does not work for most mining devices because each device has their own fork of cgminer. I think the multiple forks of cgminer make solution 1 too complex because we may have to maintain many forks of cgminer, one for each supported device. It also requires that we have access to replace the device's controller by ours.

I like solution 2 because it seems to be more aligned to things we've already discussed. It would require both a change in the consensus algorithm and the selection of a new algorithm for mining transactions.

**Comment by pedroferreira1** [comment]
I don't think we should maintain many forks of cgminer. We should stick to the forks we will use to mine transactions in our full nodes.

Solution 2 also requires some dedicated machine to mine the transactions (even though it doesn't need to be a powerful one) that would increase the servers costs. The machines we would use in the first solution would already be our mining machines, so no additional costs, I think.

I believe in solution 2 we would do all the code we will do in solution 1 (e.g. enabling NxM transaction mining) and still need to change the algorithm, so would be more work.

I think in the future we should migrate to solution 2, specially because of the feature of a client mining tx, but we should do the faster option right now in my opinion because of the costs, so I think we should go with solution 1.

**Comment by obiyankenobi** [comment]
I think we should first look for the cost of renting a GPU somewhere else. In AWS, it'd cost us almost $400. I did a quick search and there seems to be some options:
https://www.kryptex.org/en/rent-cloud-gpus
https://pro-mining.com/
https://www.nicehash.com/cpu-gpu-mining
https://vast.ai/console/create/
https://www.gpuserversrental.com/

We just need to double check if we can run our software there.

**Comment by jansegre** [comment]
> I think the multiple forks of cgminer make solution 1 too complex because we may have to maintain many forks of cgminer, one for each supported device.

I don't agree with this. We don't *have* to maintain every fork we make, all our current forks are probably obsolete now that basically all of our hashrate is coming from MM. It would probably have no impact right now if we dropped support for all 3 of the forks we already have.

And what I mean is that we are halfway on the changes needed on one of the main cgminer repos. If we have to port these changes to a cgminer fork, it probably isn't too complext.

 > I believe in solution 2 we would do all the code we will do in solution 1 (e.g. enabling NxM transaction mining) and still need to change the algorithm, so would be more work.

I don't think this would be the case if we choose an appropriate hash algorithm. The idea of using an algo, of which there aren't ASICs for, (maybe not even GPU, aside from the ASIC/GPU resistant moto, which I'm not fully convinced, but even though there are useful properties we can benefit from) is being able to use CPUs for mining. Then we could just use the CPU of the node itself (of course, running at least on a separate thread, but not necessarily on a dedicated process) and not having to  worry about stratum or any NxM with prioritizing transactions over blocks (because it would only mine transactions).

Extending a little on 2. I found an algo which may be really useful: [RandomHash](https://github.com/polyminer1/rhminer) from [PascalCoin](https://www.pascalcoin.org/). The interesting properties are: using an array of known hash functions (which we could and should change, or at least shuffle), is not memory heavy, purposely tries to be ASIC and GPU resistant (which even though I don't fully believe to be possible, it would fit very well as an anti-spam mechanism) and since we wouldn't use it on the PoW of blocks there isn't the incentive of rewards for creating ASICs, and since we could tweak it to be different enough from the exact format used by PascalCoin, we would be protected if there ever are ASICs made for PascalCoin.

These points make me think 2. is not so bad, but there is the issue of breaking the protocol, which maybe is something that we should slowly without the pressure of having heavy bills if there are delays.

> I think we should first look for the cost of renting a GPU somewhere else. In AWS, it'd cost us almost $400.

However this is also a good argument for 1. for relieving our bills sooner.

Pairing this with the ability to use the same miners for the mainnet and the testnet, this could have a significant expenditure reduction very quickly by: implementing MxN, setting up this MxN miners on two of the providers (maybe one on each is  enough) @obiyankenobi suggested. These miners would be forever ON and started and managed independently from the other nodes, and would mine TXs for all API nodes on both (mainnet/testnet) networks and possibly also mining blocks (with the background hashrate) on the testnet.

### Next actions

What is starting to be clear to me is that we need to assert these:

- What is a good estimate for the effort of making the modifications we would need on our current cgminer fork? (for option 1)
- What is a good estimate for the effort of modifying RandomHash to work on our stack (Python bindings/wrapper, maybe a built-in miner instead of a stratum miner, etc)? (for option 2)
- What are the risks/impacts of making breaking the protocol, possibly with a deadline? (for option 2)

### TL;DR

Option 1. difficulty basically boils down to how effort would be required to implement MxN on cgminer (connect to N stratum servers simultaneously) / hathor-core (send tx mining jobs to all, some or one of M miners) and setting up miners on GPU providers (or ASIC proviers if we can run our exact version of cgminer) cheaper than AWS. This depends on the effort required but seems to provide the faster and safest cut on the AWS GPU spendings.

**Comment by obiyankenobi** [comment]
> would mine TXs for all API nodes on both (mainnet/testnet) networks

I'd just add that we should only worry about mainnet for now. We can come up with solutions for the testnet, like setting `weight=1` and just using cpuminer.

**Comment by pedroferreira1** [comment]
> What is starting to be clear to me is that we need to assert these:

I think we can decide when we understand better each of the options with a better estimate of difficulty.

For me option 1 still looks better.

**Comment by msbrogli** [comment]
I agree with @obiyankenobi that the first thing is to understand how much renting a GPU costs and test it to mine transactions. But, just renting a GPU is not enough, we must be able to change the controller.

About the MxN implementation, isn't it easier to develop a Stratum Proxy that connects to multiple backends and choose the job to the client? In this case, the proxy would have a queue of transactions with high priority followed by the block. The only disadvantage is that the hash rate estimator of the stratum will not work because the client will mine blocks from one backend.

**Comment by obiyankenobi** [comment]
> In this case, the proxy would have a queue of transactions with high priority followed by the block. 

Not sure if it's worth sending blocks for this GPU for mining. It won't make any difference to the hashrate.

**Comment by jansegre** [comment]
> About the MxN implementation, isn't it easier to develop a Stratum Proxy that connects to multiple backends and choose the job to the client?

@msbrogli That kind of Stratum Proxy probably already exists for vanilla Stratum, we'd just need to find and adapt one to our modified Stratum.

> Not sure if it's worth sending blocks for this GPU for mining. It won't make any difference to the hashrate.

@obiyankenobi As far as I understand I don't think there is a way for the miner to be idle, it will continually keep mining the last jobs searching for other nonces. So we might as well leave this as is (which is mining blocks in the background).

**Comment by msbrogli** [comment]
@jansegre I think most of these proxies are focused on maximizing the profits, so maybe their code is too complex.

@obiyankenobi I completely agree. I think there may be a way to let the GPU do nothing. But I don't know if it is worth doing nothing.

**Comment by jansegre** [comment]
I looked at a few stratum proxies:

- https://github.com/slush0/stratum-mining-proxy is a getblocktemplate<->stratum proxy, not what we're looking for
- https://github.com/Stratehm/stratum-proxy most of the code is from 5 years ago, last commit is from 2017, clearly abandoned, could be a starting point though
- https://github.com/wetblanketcc/aiostratum_proxy code is in Python with asyncio, which is good, but it wasn't made for Bitcoin's stratum, it seems to have been made for Equihash coins' stratum, it isn't known whether it even works on Bitcoin's stratum, it is also unmaintained, still could be a starting point
- https://github.com/beckerich/js-stratum-proxy code is more recent (between 1 and 2 years old), seems to be geared towards what we want, this looks like the best starting point

However, from our discussion from yesterday, an idea came up to create a _tx mining API_, this looks like it would solve most of our problems with the least effort. I'll make a proposal with this.

**Comment by jansegre** [comment]
Also I have tested https://vast.ai/ and it was actually really easy to get it working. It is based on docker images, I just chose our public ccminer docker, setup the parameters to use my server, chose an available rental, and a couple minutes (probably downloading the docker image) after starting it was already mining txs. The biggest downside may be its apparent lack of team support (each user with its own login accessing the same rentals and payment methods), and also the lack of monitoring tools (logs, network status, etc).

**Comment by jansegre** [comment]
Take all of this into account, this is my latest proposal:

- Phase 1 (should start as soon as possible to gauge vast.ai's reliability): 
  - Rent GPUs on vast.ai to replace the ones currently use _as is_ (same Docker images, same params, one per API node) will probably need security groups adjustments.
- Phase 2 (_can_ start as soon as possible, has more impact on reliability than cost reduction considering 2 API nodes and 2 GPU instances):
  - Implement the tx-mining API server (we can probably use our current stratum implementation, and swap the "manager" portions for a job queue filled from HTTP requests)
  - Implement support for tx-mining API on wallets (maybe only hathor-wallet-lib would be affected)
  - Deploy tx-mining API using vast.ai GPUs, and the mobile and desktop wallets, while the two (old self-mining API and tx-mining API) are working we would be using the double number of GPU instances than needed
  - Deprecate and turn off the self-mining API (that mines txs on the API node)
- Phase 3 (somewhat open, we may choose to diverge to changing the hash algo):
  - Modify cgminer or bmminer for S9 ASIC to mine txs
  - Find vendors willing to let us run our own firmware (we know at least one)
  - Incrementally replace GPUs with ASICs (until we have acceptable reliability on ASICs availability)

**Comment by obiyankenobi** [comment]
I agree with everything. We should start tests in vast.ai ASAP. On point to look for, as discussed in the meeting earlier today, is the hashrate of the rented machines. We currently have some txs with weight 32.

**Comment by msbrogli** [comment]
I also agree with everything. We should start an RFC to design the tx-mining API. I also agree with @obiyankenobi that we should double check that the GPUs on vast.ai can timely mine transactions with weight 32.

**Comment by pedroferreira1** [comment]
Good to me

**Comment by obiyankenobi** [comment]
No need to keep this issue open anymore. This is what we've done:

1. We ended up renting GPUs on vast.ai. That served well, as it didn't require much extra work;
2. We noticed that large transactions were timing out, because the GPUs could not timely solve them;
3. We set up 2 USB ASIC miners to mine mainnet txs;
4. We are also finalizing the tx-mining service (https://github.com/HathorNetwork/tx-mining-service);

**Comment by jansegre** [comment]
I should've commented before but I had left it open because the tx-mining-service isn't in production yet. But since it is being tracked separately, and the primary concern of cost reduction is already solved, I think it's fine to leave this closed.
